Core/Src/main.c

```diff
@@ -1,181 +1,211 @@
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+uint8_t transmit_data[3] = {0x1, 0x80, 0}; // Start Bit, Control Bits for selecting CH0, Don't care bits
+uint32_t timeout_length = 250; // Timeout length in ms. If nothing is reported by then, skip.
+uint8_t receive_data[3]; // Place to put the received data.
+uint16_t adc_data; // This is the final place where ADC output will be put.
+
+const int ADC_MAX = 1023; // Maximum value 10 bits can hold. 2^10 - 1 = max int for 10 bits.
+const int COUNTER_MAX = 60000; // Maximum value of the Timer at CH1.
+
+const float DUTY_CYCLE_MIN = 0.05; // 5% is the minimum duty cycle.
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, SET); // Sets the Chip Select as High (Default state)
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1); // Start PWM for channel 1
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, RESET); // Chip Select is set to  low starting the SPI communication.
+	  HAL_SPI_TransmitReceive(&hspi1, &transmit_data, &receive_data, sizeof(receive_data), timeout_length); // Send transmit data and listen for 250ms to receive some data.
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, SET); // Chip Select is set to high to indicate the end of the SPI communication.
+
+	  // Nothing is in the first byte of received data.
+
+	  adc_data = ((uint16_t) receive_data[1] << 8) | (uint16_t) receive_data[2]; // Bit shifts the leading 8 bits by 8bits, then ORs it with the remaining 2 bits in the second index
+	  	  	  	  	  	  	  	  	  	  	  	  	  	  	  	  	  	  	  	 // Connects the separated data.
+	  // Now ADC output can be processed
+
+	  float duty_cycle = (float)(COUNTER_MAX * DUTY_CYCLE_MIN) + (float)(adc_data/ADC_MAX) * (COUNTER_MAX * DUTY_CYCLE_MIN); // First portion guarantees minimum duty cycle of DUTY_CYCLE_MIN * 100.

```

Comment:
From a readability standpoint I would suggest changing the name of `duty_cycle` just because what you are defining here isn't truly the duty cycle. The duty cycle is defined as the percentage of ON time of the PWM signal.

Also `__HAL_TIM_SET_COMPARE` does not take the "duty cycle" as a parameter, it takes a "compare value" (which you did give it but named it as the duty cycle) which it uses to compare to the current count of the timer to determine if the signal should be on or off.